src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,152 @@
+
+        protected override SyntaxList<StatementSyntax> GetBodyStatements(ForStatementSyntax forStatement)
+            => forStatement.Statement is BlockSyntax block
+                ? block.Statements
+                : SyntaxFactory.SingletonList(forStatement.Statement);
+
+        protected override bool TryGetForStatementComponents(
+            ForStatementSyntax forStatement,
+            out SyntaxToken iterationVariable, out ExpressionSyntax initializer,
+            out MemberAccessExpressionSyntax memberAccess, out ExpressionSyntax stepValue,
+            CancellationToken cancellationToken)
+        {
+            // Look for very specific forms.  Basically, only minor variations around:
+            // for (var i = 0; i < expr.Lenth; i++)
+
+            if (forStatement.Declaration != null &&
+                forStatement.Condition.IsKind(SyntaxKind.LessThanExpression) &&
+                forStatement.Incrementors.Count == 1)
+            {
+                var declaration = forStatement.Declaration;
+                if (declaration.Variables.Count == 1)
+                {
+                    var declarator = declaration.Variables[0];
+                    if (declarator.Initializer != null)
+                    {
+                        iterationVariable = declarator.Identifier;
+                        initializer = declarator.Initializer.Value;
+
+                        var binaryExpression = (BinaryExpressionSyntax)forStatement.Condition;
+
+                        // Look for:  i < expr.Length
+                        if (binaryExpression.Left is IdentifierNameSyntax identifierName &&
+                            identifierName.Identifier.ValueText == iterationVariable.ValueText &&
+                            binaryExpression.Right is MemberAccessExpressionSyntax)
+                        {
+                            memberAccess = (MemberAccessExpressionSyntax)binaryExpression.Right;
+
+                            var incrementor = forStatement.Incrementors[0];
+                            if (TryGetStepValue(iterationVariable, incrementor, out stepValue, cancellationToken))
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+
+            iterationVariable = default;
+            memberAccess = default;
+            initializer = default;
+            stepValue = default;
+            return false;
+        }
+
+        private static bool TryGetStepValue(
+            SyntaxToken iterationVariable, ExpressionSyntax incrementor,
+            out ExpressionSyntax stepValue, CancellationToken cancellationToken)
+        {
+            // support
+            //  x++
+            //  ++x
+            //  x += constant_1
+
+            ExpressionSyntax operand;
+            switch (incrementor.Kind())
+            {
+            case SyntaxKind.PostIncrementExpression:
```

Comment:
it makes me sad that there's no UnaryExpressionSyntax that would let us merge these cases

Reply:
#won'tfix